### DISCUSSION TITLE
XML Handling

### DISCUSSION THREAD

**Discussion by SiPurdy** [MarkMpn/Sql4Cds#431] [Q&A]
With the pending doom of Blocked Unmanaged Customisations (https://learn.microsoft.com/en-us/power-platform/alm/block-unmanaged-customizations) I'm looking to modify some scripts to check the IsLockdownOfUnmanagedCustomizationEnabled flag in organization.orgdborgsettings but facing little luck.

For clarification, when active layer customisation blocked, a new `<IsLockdownOfUnmanagedCustomizationEnabled>true</IsLockdownOfUnmanagedCustomizationEnabled>` element is added to the OrgSettings root element stored in organization.orgdbsettings and what I'm trying to achieve is to get the value (if it exists) to determine whether the script should exit out or attempt the updates.

In summary everything I've tried is returning the following error:

> Unhandled return type Wmhelp.XPath2.NodeIterator
> See the Execution Plan tab for details of where this error occurred

In the execution plan there is only the following:
![image](https://github.com/MarkMpn/Sql4Cds/assets/99815083/965687ec-dea1-43f3-a0c0-caa6bfe2da70)

My base query is:
``` sql
select top(1) orgdborgsettings
   , cast(orgdborgsettings as xml).value('(/OrgSettings/IsLockdownOfUnmanagedCustomizationEnabled/text())[1]','nvarchar(5)')
from organization
```

Everything I've tried has been focused on the value arguments:
- `cast(orgdborgsettings as xml).value('(/OrgSettings/IsLockdownOfUnmanagedCustomizationEnabled/text())[1]','nvarchar(5)')`
- `cast(orgdborgsettings as xml).value('(/OrgSettings/IsLockdownOfUnmanagedCustomizationEnabled)[1]','nvarchar(5)')`
- `cast(orgdborgsettings as xml).value('(//IsLockdownOfUnmanagedCustomizationEnabled)[1]','nvarchar(5)')`
- `cast(orgdborgsettings as xml).value('(//IsLockdownOfUnmanagedCustomizationEnabled/text())[1]','nvarchar(5)')`

Not sure what I've done wrong as the query is fairly similar to xml processing I've done in TSQL previously.

Any ideas.

TIA, Si

**Comment by MarkMpn** [comment]
There's some better XML support coming in the next version which will let this work with all these different variants. At the moment I think the best it can do is look for specific strings within the setting, something like:

```sql
select case when orgdborgsettings like '%<IsLockdownOfUnmanagedCustomizationEnabled>true</IsLockdownOfUnmanagedCustomizationEnabled>%' then 'Enabled' else 'Disabled' end from organization
```

**Reply by SiPurdy** [reply]
Thanks @MarkMpn, the obvious fudge that I missed.

I saw the json methods, and being from an era where xml > json > yaml I'd made the assumption the xml methods would be in there and it looked close except for the complaints on my xpath.
